_posts/2022-06-15-inline-assembly-memory-side-effects-bug.md

```diff
@@ -0,0 +1,157 @@
+new via-IR pipeline. Inline assembly blocks are never optimized in isolation when
+using that pipeline. Instead they are optimized as a part of the whole Yul input.
+
+However, the legacy code generation pipeline (which is still the default) runs the
+Yul optimizer individually on an inline assembly block if the block does not refer
+to any variables defined in the surrounding Solidity code. Consequently, memory
+writes in such inline assembly blocks are removed as well, if the written memory is
+never read from in the same assembly block, even if the written memory is accessed
+later, for example by a subsequent inline assembly block.
+
+Fortunately, the fact that the legacy code generation pipeline does not run the Yul
+optimizer at all on inline assembly blocks that do access Solidity variables,
+reduces the number of affected cases significantly. Most inline assembly blocks either
+read or write values from or to variables defined in the surrounding Solidity code,
+are entirely self-contained, or take over the program flow until the end of
+the transaction. Thereby, the bug is unlikely to occur in practice and its adverse
+effects should in most cases be easily detectable in tests. However,
+since the consequences in affected cases can be severe, we assigned a severity
+of "medium".
+
+
+In the following example, the legacy code generation pipeline with enabled optimizer will
+remove the ``mstore`` instruction and the function ``f`` will return zero:
+
+```solidity
+contract C {
+    function f() external pure returns (uint256 x) {
+        assembly {
+            mstore(0, 0x42)
+        }
+        assembly {
+            x := mload(0)
+        }
+    }
+}
+```
+
+However, if the same memory is either read again in the same inline assembly block
+or if the inline assembly block accesses any local Solidity variables, the bug is
+not present. Both is the case in the following example, so ``f`` will return
+``0x42`` as expected:
+
+```solidity
+contract C {
+    function f() external pure returns (uint256 x) {
+        assembly {
+            mstore(0, 0x42)
+            x := mload(0)
+        }
+    }
+}
+```
+
+In the following example, the first ``mstore`` will not be removed, since the written memory
+is read again by the ``return``. The second ``mstore`` on the other hand will be removed, since the memory
+it writes is never read again. In this case, this is a valid optimization and the example is not adversely
+affected by the bug. Generally, any assembly block that terminates the transaction or does not have memory side-effects that
+need to be observed afterwards is unaffected.
+
+```solidity
+contract C {
+    function f() external {
+        assembly {
+            mstore(0, 0x42)  // This write will be kept, since the return below reads from the memory.
+            mstore(32, 0x21) // This will be removed, but that is valid since the memory is never read again.
+            return(0, 32)
+        }
+    }
+}
```

Comment:
Maybe we should use `returndatacopy()` in one of these examples so that people do not think that only `mstore()` is affected? I didn't consider `returndatacopy()` myself until I saw the example from Certora.

Reply:
You mean other than the next example below :-)?

Reply:
Haha, you got me. I forgot we had that one here and now only took a quick look at the initial examples :) It's fine then.